packages/perennial-oracle/contracts/types/ChainlinkAggregator.sol

```diff
@@ -79,17 +80,97 @@ library ChainlinkAggregatorLib {
+        // lastSyncedRound is the last round in it's phase before latestRound, so we need to find where the next phase starts
         // The next phase should start at the round that is closest to but after lastSyncedRound.timestamp
-        (,,,uint256 lastSyncedRoundTimestamp,) = proxy.getRoundData(uint80(lastSyncedRoundId));
-        nextPhaseStartingRoundId = latestRound.roundId;
-        uint256 updatedAt = latestRound.timestamp;
-        // Walk back in the new phase until we dip below the lastSyncedRound.timestamp
-        while (updatedAt >= lastSyncedRoundTimestamp) {
-            nextPhaseStartingRoundId--;
-            (,,,updatedAt,) = proxy.getRoundData(uint80(nextPhaseStartingRoundId));
+        ChainlinkRound memory lastSyncedRound = getRound(self, lastSyncedRoundId);
+        uint16 phaseToSearch = lastSyncedRound.phaseId() + 1;
+        while (nextPhaseStartingRoundId == 0) {
+            nextPhaseStartingRoundId = getStartingRoundId(self, phaseToSearch, lastSyncedRound.timestamp);
+            phaseToSearch += 1;
         }
 
-        return ((lastSyncedRoundId - startingRoundId) + 1, nextPhaseStartingRoundId + 1);
+        return ((lastSyncedRoundId - startingRoundId) + 1, nextPhaseStartingRoundId);
+    }
+
+    /**
+     * @notice Returns the round ID closest to but greater than targetTimestamp for the specified phase ID
+     * @param self Chainlink Feed Aggregator to operate on
+     * @param phaseId The specific phase to fetch data for
+     * @param targetTimestamp timestamp to search for
+     * @dev Assumes the phase ends at the aggregators latestRound or earlier
+     * @return The number of rounds in the phase
+     */
+    function getStartingRoundId(ChainlinkAggregator self, uint16 phaseId, uint256 targetTimestamp)
+    internal view returns (uint256) {
+        AggregatorProxyInterface proxy = AggregatorProxyInterface(ChainlinkAggregator.unwrap(self));
+
+        (,,,uint256 startTimestamp,) = proxy.getRoundData(uint80(_aggregatorRoundIdToProxyRoundId(phaseId, 1)));
+        if (startTimestamp == 0) return 0; // Empty phase
+
+        return _search(proxy, phaseId, targetTimestamp, startTimestamp, 1);
+    }
+
+    /**
+     * Searches the given chainlink proxy for a round which has a timestamp which is as close to but greater than
+     * the `targetTimestamp`
+     * @param proxy Chainlink Proxy to search within
+     * @param phaseId Phase to search for round
+     * @param targetTimestamp Minimum timestamp value for found round
+     * @param minTimestamp Starting timestamp value
+     * @param minRoundId Starting round ID
+     */
+    function _search(AggregatorProxyInterface proxy, uint16 phaseId, uint256 targetTimestamp, uint256 minTimestamp, uint256 minRoundId) private view returns (uint256) {
+        uint256 maxRoundId = minRoundId + 1000; // Start 1000 rounds away when searching for maximum
```

Comment:
is there a limit where it won't find the solution if the phase is larger than 100 rounds?